### PR TITLE
08findnodes

### PR DESCRIPTION
**Pull request by Xliff**
There is a lot here to discuss before I expect you will want to merge. However, this commit finishes 08findnodes with no errors and also removes the existing errors from 06elements.

One thing I expect you will notice are the changes to XML::LibXML::Parser and XML::LibXML. A lot of weird things have been happening, and I had to make those changes to resolve a circular dependency.

With that said, I will let you process this at your own pace.

Take care!

### CODE REVIEW

File: lib/XML/LibXML/Document.pm [hunk 1]
```diff
@@ -9,24 +10,23 @@ use XML::LibXML::Attr;
 use XML::LibXML::Dom;
 use XML::LibXML::Enums;
 use XML::LibXML::Error;
+use XML::LibXML::Element;
 
-unit class XML::LibXML::Document is xmlDoc is repr('CStruct') does XML::LibXML::Nodish;
+multi trait_mod:<is>(Routine $r, :$aka!) is export { $r.package.^add_method($aka, $r) };
```
Comment (Xliff) [inline lib/XML/LibXML/Document.pm]: I need to test this more, but when I tried to use it, it didn't seem to work, hence the copies in all of the modules. I will see if I can remove them now that everything is running with no problems and see if they are still needed.

File: lib/XML/LibXML/Dom.pm [hunk 2]
```diff
@@ -90,34 +94,36 @@ package XML::LibXML::Dom {
...
+                # cw: -YYY- There WAS an endless loop here...          
 	            if domRemoveNsDef($tree, $tree.ns) {
-	    #           domAddNsDef($tree, $tree.ns);
-	    #       } else {
-	    #           $tree.ns = xmlCopyNamespace($tree.ns);
-	    #           domAddNsDef($tree, $tree.ns);
+                    domAddNsDef($tree, $tree.ns);
+	           } else {
+                    setObjAttr($tree, '$!ns', xmlCopyNamespace($tree.ns));
+                    domAddNsDef($tree, $tree.ns);
```
Comment (Xliff) [inline lib/XML/LibXML/Dom.pm]: Yes, hard tabs will be removed. :o)

File: lib/XML/LibXML/Element.pm [hunk 3]
```diff
@@ -10,71 +10,74 @@ use NativeCall;
 use XML::LibXML::CStructs :types;
 use XML::LibXML::Enums;
 use XML::LibXML::Node;
+use XML::LibXML::Parser;
 use XML::LibXML::Subs;
 
 also does XML::LibXML::Nodish;
 
+multi trait_mod:<is>(Routine $r, :$aka!) is export { $r.package.^add_method($aka, $r) };
+
```
Comment (Xliff) [inline lib/XML/LibXML/Element.pm]: When removed I get the following:

```
cbwood@infinity:~/projects/p6-XML-LibXML-work$ perl6 -Ilib t/06elements.t
===SORRY!=== Error while compiling /home/cbwood/projects/p6-XML-LibXML-work/lib/XML/LibXML/Element.pm (XML::LibXML::Element)
Can't use unknown trait 'is aka' in a method declaration.
at /home/cbwood/projects/p6-XML-LibXML-work/lib/XML/LibXML/Element.pm (XML::LibXML::Element):30
    expecting any of:
        rw raw hidden-from-backtrace hidden-from-USAGE
        pure default DEPRECATED inlinable nodal
        prec equiv tighter looser assoc leading_docs trailing_docs
```

**Comment by Xliff** [comment]
Note that XML::LibXML::Parser has a new parse-string that does _not_ rely on the parse() method. Without the new parse-string method, tests would not properly complete.

File: lib/XML/LibXML/Document.pm (hunk 1 above)
Comment (FROGGS) [inline lib/XML/LibXML/Document.pm]: There is an RT ticket for that...

File: lib/XML/LibXML/Element.pm (hunk 3 above)
Comment (FROGGS) [inline lib/XML/LibXML/Element.pm]: Aye, as I said, it is ticketed but the reason why it fails it not obvious.

**Pull request merged**